### PR TITLE
[DRAFT] DDS Custom Streams

### PR DESCRIPTION
**Pull request by CoderDake**
posting this up as a draft to get feedback from Ben before finishing the tests

### CODE REVIEW

File: pkg/dds/lib/src/stream_manager.dart [hunk 1]
```diff
@@ -29,8 +31,9 @@ class StreamManager {
     data, {
     DartDevelopmentServiceClient? excludedClient,
   }) {
-    if (streamListeners.containsKey(streamId)) {
-      final listeners = streamListeners[streamId]!;
+    var listeners = streamListeners[streamId];
```
Comment (bkonyi) [inline pkg/dds/lib/src/stream_manager.dart:34]: Nit: `final listeners = streamListeners[streamId] ?? customStreamListeners[streamId];`

File: pkg/dds/lib/src/stream_manager.dart [hunk 2]
```diff
@@ -138,11 +141,29 @@ class StreamManager {
     }
     dds.vmServiceClient.registerMethod(
       'streamNotify',
-      (json_rpc.Parameters parameters) {
+      (json_rpc.Parameters parameters) async {
         final streamId = parameters['streamId'].asString;
         final event =
             Event.parse(parameters['event'].asMap.cast<String, dynamic>())!;
+        final destinationStreamId =
+            event.extensionData?.data['__destinationStream']!;
```
Comment (bkonyi) [inline pkg/dds/lib/src/stream_manager.dart:149]: `'__destinationStream'` should be in a constant.

File: pkg/dds/lib/src/stream_manager.dart [hunk 3]
```diff
@@ -138,11 +141,29 @@ class StreamManager {
     }
     dds.vmServiceClient.registerMethod(
       'streamNotify',
-      (json_rpc.Parameters parameters) {
+      (json_rpc.Parameters parameters) async {
```
Comment (bkonyi) [inline pkg/dds/lib/src/stream_manager.dart:144]: This `async` isn't necessary if you're not awaiting anything in the body.

File: pkg/dds/lib/src/stream_manager.dart [hunk 4]
```diff
@@ -138,11 +141,29 @@ class StreamManager {
...
         final streamId = parameters['streamId'].asString;
         final event =
             Event.parse(parameters['event'].asMap.cast<String, dynamic>())!;
+        final destinationStreamId =
+            event.extensionData?.data['__destinationStream']!;
 
+        if (destinationStreamId != 'Extension' && destinationStreamId != null) {
```
Comment (bkonyi) [inline pkg/dds/lib/src/stream_manager.dart:151]: Use `kExtensionStream` and do the null check first.

File: sdk/lib/developer/extension.dart [hunk 5]
```diff
@@ -146,13 +146,38 @@ external bool get extensionStreamHasListener;
 /// event stream.
 ///
 /// If [extensionStreamHasListener] is false, this method is a no-op.
-void postEvent(String eventKind, Map eventData) {
+// This event will only be posted to the extension stream, if you try to post to a stream that is part of the vm service or is defined by the dart VM Embedder
+void postEvent(String eventKind, Map eventData, {String stream = 'Extension'}) {
+  // Keep protected streams in sync with `streams_` in runtime/vm/service.cc
+  // `Extension` is the only stream that should not be protected here.
+  final protectedStreams = [
```
Comment (bkonyi) [inline sdk/lib/developer/extension.dart:153]: Nit: Include the type argument for the list.

File: sdk/lib/developer/extension.dart [hunk 6]
```diff
@@ -146,13 +146,38 @@ external bool get extensionStreamHasListener;
 /// event stream.
 ///
 /// If [extensionStreamHasListener] is false, this method is a no-op.
-void postEvent(String eventKind, Map eventData) {
+// This event will only be posted to the extension stream, if you try to post to a stream that is part of the vm service or is defined by the dart VM Embedder
```
Comment (bkonyi) [inline sdk/lib/developer/extension.dart:149]: Woahhhhh Nelly, slow down! 80 character lines in the SDK ;)

File: tests/lib/developer/post_event_test.dart [hunk 7]
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2014, the Dart project authors.  Please see the AUTHORS file
```
Comment (bkonyi) [inline tests/lib/developer/post_event_test.dart:1]: Copyright (c) 2022, here and in any other new file you've added.

File: sdk/lib/developer/extension.dart [hunk 8]
```diff
@@ -146,13 +146,38 @@ external bool get extensionStreamHasListener;
...
+    'GC',
+    '_Echo',
+    'HeapSnapshot',
+    'Logging',
+    'Timeline',
+    'Profiler',
+  ];
+
+  if (protectedStreams.contains(stream)) {
+    throw new ArgumentError.value(
```
Comment (bkonyi) [inline sdk/lib/developer/extension.dart:166]: Use single quotes when possible.

File: pkg/dds/test/post_event_custom_stream_script.dart [hunk 9]
```diff
@@ -0,0 +1,9 @@
+import 'dart:developer';
```
Comment (bkonyi) [inline pkg/dds/test/post_event_custom_stream_script.dart:1]: Needs a copyright header.

File: sdk/lib/developer/extension.dart (hunk 6 above)
Comment (bkonyi) [reply sdk/lib/developer/extension.dart:149]: This documentation needs to be reworded and should explicitly document the behavior of `stream`.

File: sdk/lib/developer/extension.dart (hunk 8 above)
Comment (bkonyi) [reply sdk/lib/developer/extension.dart:166]: Also, don't use `new`.

**Comment by mit-mit** [comment]
@CoderDake is this still being worked on ?

**Comment by CoderDake** [comment]
This is done thank you for pointing out this was left open

**Pull request closed without merging**